btf/core.go

```diff
@@ -1045,13 +1050,295 @@ func coreAreMembersCompatible(localType Type, targetType Type) error {
+		}
+
+		return coreTypesMatch(lv.Type, tv.Type, behindPtr, visited)
+
+	case *FuncProto:
+		tv, ok := targetType.(*FuncProto)
+		if !ok {
+			return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+		}
+
+		if len(lv.Params) != len(tv.Params) {
+			return fmt.Errorf("function param mismatch: %w", errIncompatibleTypes)
+		}
+
+		for i, lparam := range lv.Params {
+			if err := coreTypesMatch(lparam.Type, tv.Params[i].Type, behindPtr, visited); err != nil {
+				return err
+			}
+		}
+
+		return coreTypesMatch(lv.Return, tv.Return, behindPtr, visited)
+
+	default:
+		return fmt.Errorf("unsupported type %T", localType)
+	}
+
+	return nil
+}
+
+// coreEnumsMatch checks two enums match, which is considered to be the case if the following is true:
+// - size has to match (but enum may match enum64 and vice versa)
+// - local variants have to have a match in target by symbolic name (but not numeric value)
+func coreEnumsMatch(local *Enum, target *Enum) error {
+	if local.Size != target.Size {
+		return fmt.Errorf("size mismatch between %v and %v: %w", local, target, errIncompatibleTypes)
+	}
+
+	// If there are more values in the local than the target, there must be at least one value in the local
+	// that isn't in the target, and therefor the types are incompatible.
+	if len(local.Values) > len(target.Values) {
+		return fmt.Errorf("local has more values than target: %w", errIncompatibleTypes)
+	}
+
+outer:
+	for _, lv := range local.Values {
+		for _, rv := range target.Values {
+			if coreNamesMatch(lv.Name, rv.Name) {
+				continue outer
+			}
+		}
+
+		return fmt.Errorf("no match for %v in %v: %w", lv, target, errIncompatibleTypes)
+	}
+
+	return nil
+}
+
+func coreCompositesMatch(localType, targetType composite, behindPtr bool, visited map[pair]struct{}) error {
+	if reflect.TypeOf(localType) != reflect.TypeOf(targetType) {
+		return fmt.Errorf("type mismatch between %v and %v: %w", localType, targetType, errIncompatibleTypes)
+	}
+
+	var (
+		localMembers  []Member
+		targetMembers []Member
+	)
+
+	switch lv := localType.(type) {
+	case *Struct:
+		localMembers = lv.members()
+	case *Union:
+		localMembers = lv.members()
+	default:
+		return fmt.Errorf("expected coreCompositeMatch to be called with a composite type, got %T", localType)
+	}
+
+	switch tv := targetType.(type) {
+	case *Struct:
+		targetMembers = tv.members()
+	case *Union:
+		targetMembers = tv.members()
+	default:
+		return fmt.Errorf("expected coreCompositeMatch to be called with a composite type, got %T", targetType)
+	}
+
+	for _, localMember := range localMembers {
+		matches := false
+		for _, targetMember := range targetMembers {
+			if !coreNamesMatch(localMember.Name, targetMember.Name) {
+				continue
+			}
+
+			err := coreTypesMatch(localMember.Type, targetMember.Type, behindPtr, visited)
+			if err != nil && !errors.Is(err, errIncompatibleTypes) {
```

Comment:
Why not forward all errors? Either the types are incompatible and we should return, or something unexpected happened and we should return also.

Reply:
Hmm, I have it this way because its a translation of the behavior libbpf has. Libbpf would consider these two types matching:
```
struct local {
  __u16 a
};
struct target {
  __u8 a
  __u16 a___new
};
```
You would get a `errIncompatibleTypes` when comparing `local.a` to `target.a` and then success when comparing `local.a` to `target.a___new`. Hence that we ignore `errIncompatibleTypes`. See https://github.com/libbpf/libbpf/blob/master/src/relo_core.c#L1497C9-L1508

 But, as you point out in your other comment, that implementation doesn't really make sense. The written instruction says `local members need to exist in target with the same name`. While it says `maching name` when talking about matching essential names. So I think we can conclude that this is a bug in the libbpf implementation and that we can use A) use maps for an exact name lookup. B) return `errIncompatibleTypes` as well instead of looking for other fields.